File: backend/models/rental.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type Rental struct {
	ID               primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	PropertyID       primitive.ObjectID `json:"propertyId" bson:"propertyId" binding:"required"`
	TenantID         primitive.ObjectID `json:"tenantId" bson:"tenantId" binding:"required"`
	StartDate        time.Time          `json:"startDate" bson:"startDate" binding:"required"`
	EndDate          *time.Time         `json:"endDate" bson:"endDate"`
	MonthlyRent      float64            `json:"monthlyRent" bson:"monthlyRent"`
	SecurityDeposit  float64            `json:"securityDeposit" bson:"securityDeposit"`
	Status           string             `json:"status" bson:"status"` // "active", "terminated", "expired"
	PaymentHistory   []Payment          `json:"paymentHistory" bson:"paymentHistory"`
	DueAmount        float64            `json:"dueAmount" bson:"dueAmount"`
	NextDueDate      time.Time          `json:"nextDueDate" bson:"nextDueDate"`
	Agreement        AgreementDetails   `json:"agreement" bson:"agreement"`
	CreatedAt        time.Time          `json:"createdAt" bson:"createdAt"`
	UpdatedAt        time.Time          `json:"updatedAt" bson:"updatedAt"`
}

type Payment struct {
	ID            primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	Amount        float64            `json:"amount" bson:"amount"`
	PaymentDate   time.Time          `json:"paymentDate" bson:"paymentDate"`
	PeriodMonth   int                `json:"periodMonth" bson:"periodMonth"`
	PeriodYear    int                `json:"periodYear" bson:"periodYear"`
	PaymentMethod string             `json:"paymentMethod" bson:"paymentMethod"` // "cash", "online", "cheque"
	TransactionID string             `json:"transactionId" bson:"transactionId"`
	Status        string             `json:"status" bson:"status"` // "paid", "pending", "failed"
	Remarks       string             `json:"remarks" bson:"remarks"`
	CreatedAt     time.Time          `json:"createdAt" bson:"createdAt"`
}

type AgreementDetails struct {
	Duration        int       `json:"duration" bson:"duration"` // in months
	RentIncrement   float64   `json:"rentIncrement" bson:"rentIncrement"` // yearly increment %
	NoticePeriod    int       `json:"noticePeriod" bson:"noticePeriod"` // in days
	Terms           []string  `json:"terms" bson:"terms"`
	SignedDate      time.Time `json:"signedDate" bson:"signedDate"`
	AgreementURL    string    `json:"agreementUrl" bson:"agreementUrl"`
}

type RentalRequest struct {
	PropertyID      string             `json:"propertyId" binding:"required"`
	TenantID        string             `json:"tenantId" binding:"required"`
	StartDate       time.Time          `json:"startDate" binding:"required"`
	MonthlyRent     float64            `json:"monthlyRent" binding:"required"`
	SecurityDeposit float64            `json:"securityDeposit"`
	Agreement       AgreementDetails   `json:"agreement"`
}

File: backend/models/Rental.js
const mongoose = require('mongoose');

const rentalSchema = new mongoose.Schema({
  property: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Property',
    required: true
  },
  tenant: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  landlord: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  agreement: {
    startDate: {
      type: Date,
      required: true
    },
    endDate: {
      type: Date,
      required: true
    },
    monthlyRent: {
      type: Number,
      required: true
    },
    securityDeposit: {
      type: Number,
      required: true
    },
    maintenanceCharges: {
      type: Number,
      default: 0
    },
    terms: String,
    documentUrl: String
  },
  status: {
    type: String,
    enum: ['active', 'expired', 'terminated', 'pending'],
    default: 'pending'
  },
  payments: [{
    month: {
      type: String,
      required: true // Format: "YYYY-MM"
    },
    amount: {
      type: Number,
      required: true
    },
    dueDate: {
      type: Date,
      required: true
    },
    paidDate: {
      type: Date,
      default: null
    },
    status: {
      type: String,
      enum: ['pending', 'paid', 'overdue', 'partial'],
      default: 'pending'
    },
    paymentMethod: {
      type: String,
      enum: ['cash', 'upi', 'bank_transfer', 'cheque', 'online'],
      default: null
    },
    transactionId: String,
    late_fee: {
      type: Number,
      default: 0
    },
    notes: String
  }],
  totalDue: {
    type: Number,
    default: 0
  },
  lastPaymentDate: {
    type: Date,
    default: null
  },
  notices: [{
    type: {
      type: String,
      enum: ['rent_due', 'late_payment', 'maintenance', 'general', 'termination']
    },
    message: String,
    date: {
      type: Date,
      default: Date.now
    },
    acknowledged: {
      type: Boolean,
      default: false
    }
  }],
  maintenance: [{
    issue: String,
    description: String,
    reportedDate: {
      type: Date,
      default: Date.now
    },
    status: {
      type: String,
      enum: ['reported', 'in_progress', 'completed', 'rejected'],
      default: 'reported'
    },
    assignedTo: String,
    completedDate: Date,
    cost: Number,
    images: [String]
  }]
}, {
  timestamps: true
});

// Index for efficient queries
rentalSchema.index({ tenant: 1, status: 1 });
rentalSchema.index({ property: 1 });
rentalSchema.index({ 'payments.status': 1 });
rentalSchema.index({ 'agreement.endDate': 1 });

// Calculate total due before saving
rentalSchema.pre('save', function(next) {
  const pendingPayments = this.payments.filter(payment => 
    payment.status === 'pending' || payment.status === 'overdue' || payment.status === 'partial'
  );
  
  this.totalDue = pendingPayments.reduce((total, payment) => {
    return total + payment.amount + (payment.late_fee || 0);
  }, 0);
  
  next();
});

module.exports = mongoose.model('Rental', rentalSchema);